csds-client/client/v2/client.go

```diff
@@ -0,0 +1,290 @@
+		return err
+	}
+
+	c.nodeMatcher = nodematchers
+
+	// check if required fields exist in nodematcher
+	switch c.opts.Platform {
+	case "gcp":
+		keys := []string{"TRAFFICDIRECTOR_GCP_PROJECT_NUMBER", "TRAFFICDIRECTOR_NETWORK_NAME"}
+		for _, key := range keys {
+			if value := client.GetValueByKeyFromNodeMatcher(c.nodeMatcher, key); value == "" {
+				return fmt.Errorf("missing field %v in NodeMatcher", key)
+			}
+		}
+	default:
+		return fmt.Errorf("%s platform is not supported, list of supported platforms: gcp", c.opts.Platform)
+	}
+
+	return nil
+}
+
+// connWithAuth connects to uri with authentication
+func (c *ClientV2) connWithAuth() error {
+	var scope string
+	switch c.opts.AuthnMode {
+	case "jwt":
+		if c.opts.Jwt == "" {
+			return errors.New("missing jwt file")
+		}
+		switch c.opts.Platform {
+		case "gcp":
+			scope = "https://www.googleapis.com/auth/cloud-platform"
+			pool, err := x509.SystemCertPool()
+			if err != nil {
+				return err
+			}
+			creds := credentials.NewClientTLSFromCert(pool, "")
+			perRPC, err := oauth.NewServiceAccountFromFile(c.opts.Jwt, scope)
+			if err != nil {
+				return err
+			}
+
+			c.clientConn, err = grpc.Dial(c.opts.Uri, grpc.WithTransportCredentials(creds), grpc.WithPerRPCCredentials(perRPC))
+			if err != nil {
+				return err
+			}
+			return nil
+		default:
+			return fmt.Errorf("%s platform is not supported, list of supported platforms: gcp", c.opts.Platform)
+		}
+	case "auto":
+		switch c.opts.Platform {
+		case "gcp":
+			scope = "https://www.googleapis.com/auth/cloud-platform"
+			pool, err := x509.SystemCertPool()
+			if err != nil {
+				return err
+			}
+			creds := credentials.NewClientTLSFromCert(pool, "")
+			perRPC, err := oauth.NewApplicationDefault(context.Background(), scope) // Application Default Credentials (ADC)
+			if err != nil {
+				return err
+			}
+
+			// parse GCP project number as header for authentication
+			var key string
+			switch c.opts.Uri {
+			case "trafficdirector.googleapis.com:443":
+				key = "TRAFFICDIRECTOR_GCP_PROJECT_NUMBER"
+			}
+			if projectNum := client.GetValueByKeyFromNodeMatcher(c.nodeMatcher, key); projectNum != "" {
+				c.metadata = metadata.Pairs("x-goog-user-project", projectNum)
+			}
+
+			c.clientConn, err = grpc.Dial(c.opts.Uri, grpc.WithTransportCredentials(creds), grpc.WithPerRPCCredentials(perRPC))
+			if err != nil {
+				return err
+			}
+			return nil
+		default:
+			return errors.New("auto authentication mode for this platform is not supported. Please use jwt_file instead")
+		}
+	default:
+		return errors.New("invalid authn_mode")
+	}
+}
+
+// New creates a new client with v2 api version
+func New(option client.ClientOptions) (*ClientV2, error) {
+	c := &ClientV2{
+		opts: option,
+	}
+	if c.opts.Platform != "gcp" {
+		return nil, fmt.Errorf("%s platform is not supported, list of supported platforms: gcp", c.opts.Platform)
+	}
+	if c.opts.ApiVersion != "v2" {
```

Comment:
We don't need to check it anymore, since creation of this client assumes v2 and v2 only.

Reply:
Since the code has been refactored and may be reused in future, is there any possible that people call `v2.New()` in `main` without checking the api version of v2? Or maybe I can force the `c.opts.Platform` to be v2 in `v2.New()` by assigning it a default value